tracing-subscriber/src/layer.rs

```diff
@@ -0,0 +1,255 @@
+use tracing_core::{
+    metadata::Metadata,
+    span,
+    subscriber::{Interest, Subscriber},
+    Event,
+};
+
+use crate::filter;
+use std::any::TypeId;
+
+pub trait Layer<S>: 'static {
+    /// Registers a new callsite with this layer, returning whether or not
+    /// the subscriber is interested in being notified about the callsite.
+    ///
+    /// This function is provided with the `Interest` returned by the wrapped
+    /// subscriber. The layer may then choose to return
+    fn register_callsite(&self, _metadata: &'static Metadata<'static>, prev: Interest) -> Interest {
+        prev
+    }
+    fn enabled(&self, _metadata: &Metadata, prev: bool) -> bool {
+        prev
+    }
+    fn new_span(&self, _attrs: &span::Attributes, _id: &span::Id) {}
```

Comment:
are these supposed to be empty impls? I would imagine the user would have to forward the call?

Reply:
`Layered` implements _the `Subscriber` trait_ by forwarding the call; this is why `Layer::new_span`gets passed the `span::Id` _already_ returned by the root `new_span` call and this function returns nothing. The `Layer` trait differs from the `Subscriber` trait in order to ensure that `Layer`s cannot override span IDs, etc